src/lib-imap/imap-capability-list.h

```diff
@@ -0,0 +1,83 @@
+#ifndef IMAP_CAPABILITY_LIST_H
+#define IMAP_CAPABILITY_LIST_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "lib.h"
+#include "array.h"
+
+/* bitflags to represent different conditions of
+   the current imap session */
+typedef enum {
+	CAP_NEVER       = 0,         /* never show this capability */
+	CAP_PREAUTH     = (1 << 0),  /* show this cap before login */
+	CAP_POSTAUTH    = (1 << 1),  /* show this cap after login */
+	CAP_TLSACTIVE   = (1 << 2),  /* show this cap if tls is active */
+	CAP_TLSINACTIVE = (1 << 3),  /* show this cap if tls is inactive */
+	CAP_SSLACTIVE   = (1 << 4),  /* show this cap if ssl is active */
+	CAP_SSLINACTIVE = (1 << 5),  /* show this cap if ssl is inactive */
```

Comment:
Why is there both TLS and SSL? Also it's wrong to call TLS connection started with STARTTLS command as "TLS" and TLS connection started with the imaps port "SSL". I don't think there's any need to separate them anyway? Simpler to just have SSLACTIVE and SSL_INACTIVE? (Dovecot generally calls all of these "ssl" rather than "tls", although maybe I should have made them all "tls" since the beginning. But too late to change now.)

Also, instead of 1<<n use BIT(n)

Reply:
I have adjusted the code to use BIT(n), and I have removed the SSL_ACTIVE/INACTIVE and retained the IMAP_CAP_VISIBILITY_TLS_ACTIVE and IMAP_CAP_VISIBILITY_TLS_INACTIVE.